{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### @Mike \n",
    "# Can you add a few sentences about how you got the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary modules\n",
    "import os\n",
    "import glob\n",
    "\n",
    "import pandas as pd\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "import plotly.graph_objects as go # for interactive plot with human defined labels\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>DNA</th>\n",
       "      <th>RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TOBG-MED-1076_1116</td>\n",
       "      <td>0.714860</td>\n",
       "      <td>4.037256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TOBG-MED-1076_1131</td>\n",
       "      <td>7.727037</td>\n",
       "      <td>5.454919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TOBG-MED-1076_1151</td>\n",
       "      <td>2.859440</td>\n",
       "      <td>15.572275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TOBG-MED-1076_1195</td>\n",
       "      <td>8.813051</td>\n",
       "      <td>12.379700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene       DNA        RNA\n",
       "0  TOBG-MED-1076_1101  3.578633  12.992625\n",
       "1  TOBG-MED-1076_1116  0.714860   4.037256\n",
       "2  TOBG-MED-1076_1131  7.727037   5.454919\n",
       "3  TOBG-MED-1076_1151  2.859440  15.572275\n",
       "4  TOBG-MED-1076_1195  8.813051  12.379700"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load necessary data\n",
    "path = \"/Users/ZXue/src/speeding-up-sci-correlation/\"\n",
    "df = pd.read_table(os.path.join(path, \"TARA_025-merged-only-those-in-both-full-TPM.tsv\"))\n",
    "df.head() # a look at the sample\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>DNA</th>\n",
       "      <th>RNA</th>\n",
       "      <th>KO_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "      <td>K01940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TOBG-MED-1076_1116</td>\n",
       "      <td>0.714860</td>\n",
       "      <td>4.037256</td>\n",
       "      <td>K00962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TOBG-MED-1076_1131</td>\n",
       "      <td>7.727037</td>\n",
       "      <td>5.454919</td>\n",
       "      <td>K01696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TOBG-MED-1076_1151</td>\n",
       "      <td>2.859440</td>\n",
       "      <td>15.572275</td>\n",
       "      <td>K02990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>TOBG-MED-1076_1298</td>\n",
       "      <td>2.923141</td>\n",
       "      <td>4.295652</td>\n",
       "      <td>K00347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24039</td>\n",
       "      <td>TOBG-MED-831_959</td>\n",
       "      <td>1.353264</td>\n",
       "      <td>7.915675</td>\n",
       "      <td>K00005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24040</td>\n",
       "      <td>TOBG-MED-831_960</td>\n",
       "      <td>8.061755</td>\n",
       "      <td>34.266264</td>\n",
       "      <td>K03696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24041</td>\n",
       "      <td>TOBG-MED-831_978</td>\n",
       "      <td>0.950462</td>\n",
       "      <td>1.150253</td>\n",
       "      <td>K03455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24042</td>\n",
       "      <td>TOBG-MED-831_98</td>\n",
       "      <td>0.479963</td>\n",
       "      <td>3.630331</td>\n",
       "      <td>K08309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24043</td>\n",
       "      <td>TOBG-MED-831_984</td>\n",
       "      <td>1.385115</td>\n",
       "      <td>3.352542</td>\n",
       "      <td>K05592</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13165 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene       DNA        RNA   KO_ID\n",
       "0      TOBG-MED-1076_1101  3.578633  12.992625  K01940\n",
       "1      TOBG-MED-1076_1116  0.714860   4.037256  K00962\n",
       "2      TOBG-MED-1076_1131  7.727037   5.454919  K01696\n",
       "3      TOBG-MED-1076_1151  2.859440  15.572275  K02990\n",
       "6      TOBG-MED-1076_1298  2.923141   4.295652  K00347\n",
       "...                   ...       ...        ...     ...\n",
       "24039    TOBG-MED-831_959  1.353264   7.915675  K00005\n",
       "24040    TOBG-MED-831_960  8.061755  34.266264  K03696\n",
       "24041    TOBG-MED-831_978  0.950462   1.150253  K03455\n",
       "24042     TOBG-MED-831_98  0.479963   3.630331  K08309\n",
       "24043    TOBG-MED-831_984  1.385115   3.352542  K05592\n",
       "\n",
       "[13165 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add KO ID to df\n",
    "df_anno = pd.read_table(os.path.join(path, \"Our-Kegg-annotations.tsv\"))\n",
    "df_anno.head()\n",
    "\n",
    "df = df.merge(df_anno, on='Gene')\n",
    "# remove genes that are not annotated with a KO ID\n",
    "df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>DNA</th>\n",
       "      <th>RNA</th>\n",
       "      <th>KO_ID</th>\n",
       "      <th>Category1</th>\n",
       "      <th>Category2</th>\n",
       "      <th>Category3</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "      <td>K01940</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Overview</td>\n",
       "      <td>01230 Biosynthesis of amino acids [PATH:ko01230]</td>\n",
       "      <td>argG, ASS1; argininosuccinate synthase [EC:6....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "      <td>K01940</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>00250 Alanine, aspartate and glutamate metabol...</td>\n",
       "      <td>argG, ASS1; argininosuccinate synthase [EC:6....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "      <td>K01940</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>00220 Arginine biosynthesis [PATH:ko00220]</td>\n",
       "      <td>argG, ASS1; argininosuccinate synthase [EC:6....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TOBG-MED-1076_1101</td>\n",
       "      <td>3.578633</td>\n",
       "      <td>12.992625</td>\n",
       "      <td>K01940</td>\n",
       "      <td>Human Diseases</td>\n",
       "      <td>Cardiovascular diseases</td>\n",
       "      <td>05418 Fluid shear stress and atherosclerosis [...</td>\n",
       "      <td>argG, ASS1; argininosuccinate synthase [EC:6....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TOBG-MED-590_1583</td>\n",
       "      <td>23.569356</td>\n",
       "      <td>25.271020</td>\n",
       "      <td>K01940</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Overview</td>\n",
       "      <td>01230 Biosynthesis of amino acids [PATH:ko01230]</td>\n",
       "      <td>argG, ASS1; argininosuccinate synthase [EC:6....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene        DNA        RNA   KO_ID       Category1  \\\n",
       "0  TOBG-MED-1076_1101   3.578633  12.992625  K01940      Metabolism   \n",
       "1  TOBG-MED-1076_1101   3.578633  12.992625  K01940      Metabolism   \n",
       "2  TOBG-MED-1076_1101   3.578633  12.992625  K01940      Metabolism   \n",
       "3  TOBG-MED-1076_1101   3.578633  12.992625  K01940  Human Diseases   \n",
       "4   TOBG-MED-590_1583  23.569356  25.271020  K01940      Metabolism   \n",
       "\n",
       "                 Category2                                          Category3  \\\n",
       "0                 Overview   01230 Biosynthesis of amino acids [PATH:ko01230]   \n",
       "1    Amino acid metabolism  00250 Alanine, aspartate and glutamate metabol...   \n",
       "2    Amino acid metabolism         00220 Arginine biosynthesis [PATH:ko00220]   \n",
       "3  Cardiovascular diseases  05418 Fluid shear stress and atherosclerosis [...   \n",
       "4                 Overview   01230 Biosynthesis of amino acids [PATH:ko01230]   \n",
       "\n",
       "                                         description  \n",
       "0   argG, ASS1; argininosuccinate synthase [EC:6....  \n",
       "1   argG, ASS1; argininosuccinate synthase [EC:6....  \n",
       "2   argG, ASS1; argininosuccinate synthase [EC:6....  \n",
       "3   argG, ASS1; argininosuccinate synthase [EC:6....  \n",
       "4   argG, ASS1; argininosuccinate synthase [EC:6....  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add pathway info to df\n",
    "df_path = pd.read_table(os.path.join(path, \"Kegg-Orthology-table.tsv\"))\n",
    "df_path.head()\n",
    "\n",
    "df = df.merge(df_path, on='KO_ID')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Make the figure with scatter dots\n",
    "tracedot = go.Scatter(x=df['DNA'], # User should change this based on dataframe columns\n",
    "                      y=df['RNA'], # User defined \n",
    "                      mode='markers', # for scatter plot\n",
    "                      text=df['KO_ID'], # hover text\n",
    "                      #color=df['Category1'],\n",
    "                      name='Data') # User defined\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'smf' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-a682ae15fbe3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Initialise and fit linear regression model using `statsmodels`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m### tutorial https://towardsdatascience.com/introduction-to-linear-regression-in-python-c12a072bedf0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msmf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mols\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'RNA ~ DNA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# get linear regression parameters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'smf' is not defined"
     ]
    }
   ],
   "source": [
    "# Initialise and fit linear regression model using `statsmodels`\n",
    "### tutorial https://towardsdatascience.com/introduction-to-linear-regression-in-python-c12a072bedf0\n",
    "model = smf.ols('RNA ~ DNA', data=df)\n",
    "model = model.fit()\n",
    "# get linear regression parameters\n",
    "model.params\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.02598684167801979"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get linear regression R^2 \n",
    "model.rsquared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Intercept     7.292660e-07\n",
       "DNA          8.397350e-103\n",
       "dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get linear regression p-value\n",
    "model.pvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# linear regression equation\n",
    "max_val = df['DNA'].max()\n",
    "df2 = [[0, model.params[0]],[max_val, model.params[1]*max_val+model.params[0]]]\n",
    "df2 = pd.DataFrame(df2, columns = ['xi', 'y']) \n",
    " \n",
    "\n",
    "# Make the line graph \n",
    "trace_ln = go.Scatter(\n",
    "                  x=df2['xi'],\n",
    "                  y=df2['y'],\n",
    "                  mode='lines',\n",
    "                  name='Fit'\n",
    "                  )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'model' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-adbe7a76d624>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Make the text string indicating the linear regression model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0meq_txt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"R\\N{SUPERSCRIPT TWO} = {:.3f}, \\y = {:.3f}x+{:.3f}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrsquared\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m annotation = go.layout.Annotation(\n",
      "\u001b[0;31mNameError\u001b[0m: name 'model' is not defined"
     ]
    }
   ],
   "source": [
    "# Make the text string indicating the linear regression model\n",
    "\n",
    "eq_txt = \"R\\N{SUPERSCRIPT TWO} = {:.3f}, \\y = {:.3f}x+{:.3f}\".format(model.rsquared, model.params[1], model.params[0])\n",
    "\n",
    "annotation = go.layout.Annotation(\n",
    "                  x=df['DNA'].max(),\n",
    "                  y=df['RNA'].max(),\n",
    "                  text=eq_txt,\n",
    "                  showarrow=False,\n",
    "                  font=go.Font(size=16)\n",
    "                  )\n",
    "layout = go.Layout(annotations=[annotation])\n",
    "\n",
    "\n",
    "# Tie everything together to a figure\n",
    "fig = go.Figure([tracedot,trace_ln], layout=layout)\n",
    "\n",
    "\n",
    "# Edit the figure layout \n",
    "fig.update_layout(title='Gene correlations between metagenomics and metatranscriptomic resutls',\n",
    "                  xaxis_title='Count (Metagenomics)',\n",
    "                  yaxis_title='Count (Metatranscriptomics)')\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### go trace plotting can not set color to dots by category \n",
    "## so I am now trying to set color with px plotting\n",
    "## and add line and text through adding trace\n",
    "figdot = px.scatter(df, x=\"DNA\", y=\"RNA\", color=\"Category2\", hover_data=['KO_ID'])\n",
    "\n",
    "# a \"dummy\" scatter plot trace that is used to carry the text \n",
    "trace_txt = go.Scatter(   \n",
    "    x=[df['DNA'].mean()], ##########\n",
    "    y=[df['RNA'].max()],\n",
    "    mode='markers+text',\n",
    "    text=eq_txt,\n",
    "    marker=dict(size = 1),\n",
    "    name='linear regression model',\n",
    "    textposition='bottom center'\n",
    ")\n",
    "\n",
    "# add a trace to represent 1:1 regression (y=x)\n",
    "df3 = [[0, 0],[max_val, max_val]]\n",
    "df3 = pd.DataFrame(df3, columns = ['xi', 'y']) \n",
    "trace_yx = go.Scatter(\n",
    "                  x=df3['xi'],\n",
    "                  y=df3['y'],\n",
    "                  mode='lines',\n",
    "                  line = dict(color='grey', width=4, dash='dot'),\n",
    "                  name='1:1 line'\n",
    "                  )\n",
    "\n",
    "\n",
    "figdot.add_trace(trace_ln).add_trace(trace_txt).add_trace(trace_yx)\n",
    "#go.Figure(figdot, layout=layout) # why is the text in layout settings not showing up?!\n",
    "figdot.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>DNA</th>\n",
       "      <th>RNA</th>\n",
       "      <th>KO_ID</th>\n",
       "      <th>Category1</th>\n",
       "      <th>Category2</th>\n",
       "      <th>Category3</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>9592</td>\n",
       "      <td>TOBG-MED-587_690</td>\n",
       "      <td>11.446196</td>\n",
       "      <td>37.039658</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Membrane transport</td>\n",
       "      <td>02010 ABC transporters [PATH:ko02010]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9593</td>\n",
       "      <td>TOBG-MED-587_690</td>\n",
       "      <td>11.446196</td>\n",
       "      <td>37.039658</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Signal transduction</td>\n",
       "      <td>02020 Two-component system [PATH:ko02020]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9594</td>\n",
       "      <td>TOBG-MED-587_690</td>\n",
       "      <td>11.446196</td>\n",
       "      <td>37.039658</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Human Diseases</td>\n",
       "      <td>Infectious diseases</td>\n",
       "      <td>05152 Tuberculosis [PATH:ko05152]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9595</td>\n",
       "      <td>TOBG-MED-591_1040</td>\n",
       "      <td>94.927042</td>\n",
       "      <td>7.677695</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Membrane transport</td>\n",
       "      <td>02010 ABC transporters [PATH:ko02010]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9596</td>\n",
       "      <td>TOBG-MED-591_1040</td>\n",
       "      <td>94.927042</td>\n",
       "      <td>7.677695</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Signal transduction</td>\n",
       "      <td>02020 Two-component system [PATH:ko02020]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9662</td>\n",
       "      <td>TOBG-MED-831_13</td>\n",
       "      <td>0.502104</td>\n",
       "      <td>11.241493</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Signal transduction</td>\n",
       "      <td>02020 Two-component system [PATH:ko02020]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9663</td>\n",
       "      <td>TOBG-MED-831_13</td>\n",
       "      <td>0.502104</td>\n",
       "      <td>11.241493</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Human Diseases</td>\n",
       "      <td>Infectious diseases</td>\n",
       "      <td>05152 Tuberculosis [PATH:ko05152]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9664</td>\n",
       "      <td>TOBG-MED-831_64</td>\n",
       "      <td>28.375790</td>\n",
       "      <td>217.281645</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Membrane transport</td>\n",
       "      <td>02010 ABC transporters [PATH:ko02010]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9665</td>\n",
       "      <td>TOBG-MED-831_64</td>\n",
       "      <td>28.375790</td>\n",
       "      <td>217.281645</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Environmental Information Processing</td>\n",
       "      <td>Signal transduction</td>\n",
       "      <td>02020 Two-component system [PATH:ko02020]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9666</td>\n",
       "      <td>TOBG-MED-831_64</td>\n",
       "      <td>28.375790</td>\n",
       "      <td>217.281645</td>\n",
       "      <td>K02040</td>\n",
       "      <td>Human Diseases</td>\n",
       "      <td>Infectious diseases</td>\n",
       "      <td>05152 Tuberculosis [PATH:ko05152]</td>\n",
       "      <td>pstS; phosphate transport system substrate-bi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene        DNA         RNA   KO_ID  \\\n",
       "9592   TOBG-MED-587_690  11.446196   37.039658  K02040   \n",
       "9593   TOBG-MED-587_690  11.446196   37.039658  K02040   \n",
       "9594   TOBG-MED-587_690  11.446196   37.039658  K02040   \n",
       "9595  TOBG-MED-591_1040  94.927042    7.677695  K02040   \n",
       "9596  TOBG-MED-591_1040  94.927042    7.677695  K02040   \n",
       "...                 ...        ...         ...     ...   \n",
       "9662    TOBG-MED-831_13   0.502104   11.241493  K02040   \n",
       "9663    TOBG-MED-831_13   0.502104   11.241493  K02040   \n",
       "9664    TOBG-MED-831_64  28.375790  217.281645  K02040   \n",
       "9665    TOBG-MED-831_64  28.375790  217.281645  K02040   \n",
       "9666    TOBG-MED-831_64  28.375790  217.281645  K02040   \n",
       "\n",
       "                                 Category1            Category2  \\\n",
       "9592  Environmental Information Processing   Membrane transport   \n",
       "9593  Environmental Information Processing  Signal transduction   \n",
       "9594                        Human Diseases  Infectious diseases   \n",
       "9595  Environmental Information Processing   Membrane transport   \n",
       "9596  Environmental Information Processing  Signal transduction   \n",
       "...                                    ...                  ...   \n",
       "9662  Environmental Information Processing  Signal transduction   \n",
       "9663                        Human Diseases  Infectious diseases   \n",
       "9664  Environmental Information Processing   Membrane transport   \n",
       "9665  Environmental Information Processing  Signal transduction   \n",
       "9666                        Human Diseases  Infectious diseases   \n",
       "\n",
       "                                      Category3  \\\n",
       "9592      02010 ABC transporters [PATH:ko02010]   \n",
       "9593  02020 Two-component system [PATH:ko02020]   \n",
       "9594          05152 Tuberculosis [PATH:ko05152]   \n",
       "9595      02010 ABC transporters [PATH:ko02010]   \n",
       "9596  02020 Two-component system [PATH:ko02020]   \n",
       "...                                         ...   \n",
       "9662  02020 Two-component system [PATH:ko02020]   \n",
       "9663          05152 Tuberculosis [PATH:ko05152]   \n",
       "9664      02010 ABC transporters [PATH:ko02010]   \n",
       "9665  02020 Two-component system [PATH:ko02020]   \n",
       "9666          05152 Tuberculosis [PATH:ko05152]   \n",
       "\n",
       "                                            description  \n",
       "9592   pstS; phosphate transport system substrate-bi...  \n",
       "9593   pstS; phosphate transport system substrate-bi...  \n",
       "9594   pstS; phosphate transport system substrate-bi...  \n",
       "9595   pstS; phosphate transport system substrate-bi...  \n",
       "9596   pstS; phosphate transport system substrate-bi...  \n",
       "...                                                 ...  \n",
       "9662   pstS; phosphate transport system substrate-bi...  \n",
       "9663   pstS; phosphate transport system substrate-bi...  \n",
       "9664   pstS; phosphate transport system substrate-bi...  \n",
       "9665   pstS; phosphate transport system substrate-bi...  \n",
       "9666   pstS; phosphate transport system substrate-bi...  \n",
       "\n",
       "[75 rows x 8 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''An example of interesting results: K01574 was expressed at very high level in metatrancriptomic results but its \"copy number\"\n",
    "is relatively low by metagenomic measurements. '''\n",
    "\n",
    "'''This KO ID is linked to both carbohydrate and lipid metabolism, suggesting robust microbial activities\n",
    "within in the sampled Tara oceam specimen.'''\n",
    "\n",
    "df[df['KO_ID']  == 'K02040']\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42.77276000869718"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## To make a web-based application using plotly?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
